docs/design/eth-subscribe.md

```diff
@@ -0,0 +1,366 @@
+|     32609     |                                                 Invalid subscription id                                                  | Occurs when user tries to unsubscribe from a non-existing subscription or one that is not associated with the active socket connection. |
+|     32610     |                                    Invalid subscription parameters: {detailed reason}                                    |               Occurs when there is an error due to the provided subscription parameters, i.e. the address does not exist.               |
+|     32611     |                                                   Subscription aborted                                                   |                          Occurs when the subscription is aborted by the relay, i.e. a maximum TTL is reached.                           |
+
+## Rate Limits
+
+Active subscriptions should be limited in some way. We should create a flexible and configurable system for limiting subscriptions on a global and an IP level. This should be done by extending the existing `RateLimit` class as needed.
+
+1. Add a configurable global limit on the total active subscriptions.
+2. Add a configurable limit on the total active subscriptions per IP.
+3. Add a TTL to subscriptions, i. e. automatically terminate a subscription after a configurable time period.
+4. Make it possible to disable all of those limits if needed.
+
+
+## Metric Capturing
+
+Capture metrics for the following:
+
+1. Log every call to `eth_subscribe` and `eth_unsubscribe`, as a total amount and broken down by IP.
+2. The duration of active subscriptions.
+3. The total amount of requests to the mirror node per subscription.
+4. The amount of requests to the mirror node per subscription that have returned non-null data.
+
+## Tests
+
+The following test cases should be covered but additional tests would be welcome.
+
+1. Connecting to the ws-server works.
+2. Subscribing to the `logs` event with all possible filter combinations - no filters, by address, by topics, multiple addresses, etc..
+3. Unsubscribing from an active subscription.
+4. Receiving the correct log data after an async process generates logs for that match/don't match the subscription filters.
+5. Multiple active subscriptions and each of them receives their appropriate log data.
+6. Exceeding the global subscription limit.
+7. Exceeding the IP subscription limit.
+8. Unsubscribing due to automatic subscription termination. 
+9. E2E test using popular libraries (`ethers.js WebSocketProvider`) that include - connecting, subscribing, receiving data, unsubscribing
+
+## Non-Functional Requirements
+
+Users should be required to renew their subscription in the case of an error or if the subscription was closed from the relay side. If for some reason the relay is restarted it should not automatically try to restore all previous connections and subscriptions.
+
+## Open Questions
+
+1. How is the deployment going to be handled?
```

Comment:
Do we have answers yet?
Initially we'll want the same container instance to serve these with a future option to split out as needed.
This means there should be an item to ensure container start manages both processes.